### PR TITLE
tools: emscripten

### PR DESCRIPTION
**Pull request by tsvisabo**
Add emscripten toolchain,
currently builds for x64 linux,
once the PR is merged, and if not resolved by then, should add an issue to
solve absolute path creation in autodependency files (.d files) generated bazel
so we want have to pass 2 path (workspace dir and external) as `define`

### CODE REVIEW

File: tools/emscripten/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# IOTA toolchains - emscripten
+
+Transpile command:
+```
+bazel build --define=workspace=$(bazel info workspace)  --define=output_base=$(bazel info output_base) --crosstool_top=//tools/emscripten:emscripten --config='emscripten'  //tests:hi.js
```
Comment (thibault-martinez) [inline tools/emscripten/README.md]: `crosstool_top=//tools/emscripten:emscripten` is already defined in `--config='emscripten'` so no need to add it here

File: .bazelrc [hunk 2]
```diff
@@ -20,3 +20,7 @@ build:x86_64 --host_crosstool_top=@bazel_tools//tools/cpp:toolchain
 build:esp32_64 --crosstool_top=//tools/xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0:toolchain
 build:esp32_64 --cpu=xtensa
 build:esp32_64 --host_crosstool_top=@bazel_tools//tools/cpp:toolchain
+
+build:emscripten --crosstool_top=//tools/emscripten:emscripten
```
Comment (thibault-martinez) [inline .bazelrc]: Can we call the actual toolchain `toolchain` for consistency with the other toolchains ?

File: tests/BUILD [hunk 3]
```diff
@@ -11,3 +11,16 @@ cc_library(
     #    includes = [".",],
     visibility = ["//visibility:public"],
 )
+
+load("//tools/emscripten:defs.bzl", "emcc_binary")
```
Comment (thibault-martinez) [inline tests/BUILD]: Can you put the load at the top of the file ?

File: toolchains.bzl [hunk 4]
```diff
@@ -45,3 +45,25 @@ def setup_toolchains():
         strip_prefix = "xtensa-esp32-elf",
         url = "https://dl.espressif.com/dl/xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz",
     )
+
+    #emscripten
+    http_archive(
```
Comment (thibault-martinez) [inline toolchains.bzl]: Add sha256 to avoid bazel warning

File: toolchains.bzl [hunk 5]
```diff
@@ -45,3 +45,25 @@ def setup_toolchains():
...
+        name = "emscripten_toolchain",
+        build_file = "//tools/emscripten:emscripten-toolchain.BUILD",
+        strip_prefix = "emscripten-1.38.31",
+        url = "https://github.com/kripken/emscripten/archive/1.38.31.tar.gz",
+    )
+
+    http_archive(
```
Comment (thibault-martinez) [inline toolchains.bzl]: Add sha256 to avoid bazel warning

File: toolchains.bzl (hunk 4 above)
Comment (thibault-martinez) [reply toolchains.bzl]: `c87e42cb6a104094e7daf2b7e61ac835f83674ac0168f533455838a1129cc764`

File: toolchains.bzl (hunk 5 above)
Comment (thibault-martinez) [reply toolchains.bzl]: `a0c2f2c5a897577f40af0fdf68dcf3cf65557ff20c081df26678c066a4fed4b1`

File: tools/emscripten/emscripten_cache/asmjs/generated_struct_info.json [hunk 6]
```diff
@@ -0,0 +1 @@
...ID_VARIANT_NCS":0,"ESPIPE":29,"AUDIO_S32MSB":36896,"EMSCRIPTEN_EVENT_WHEEL":9,"SDL_AUDIO_ALLOW_CHANNELS_CHANGE":4,"_SC_XOPEN_REALTIME":130,"EL2HLT":51,"TCSETAW":21511,"EMSCRIPTEN_FETCH_NO_DOWNLOAD":32,"EAI_NONAME":-2,"_PC_PIPE_BUF":5,"EOWNERDEAD":130,"EROFS":30,"TCSETAF":21512,"ECONNREFUSED":111,"_SC_2_PBS_ACCOUNTING":169,"EMSCRIPTEN_EVENT_FOCUS":13,"AUDIO_S16SYS":32784,"ENETDOWN":100,"ENOEXEC":8,"EMSCRIPTEN_EVENT_TARGET_DOCUMENT":1,"ENOSPC":28,"EBADF":9,"EBADE":52,"EMSCRIPTEN_EVENT_DBLCLICK":7,"EDOTDOT":73,"_SC_THREAD_ATTR_STACKSIZE":78,"EBADFD":77,"O_ACCMODE":2097155,"EBADR":53,"_SC_2_VERSION":46,"S_IFCHR":8192,"SDL_PATCHLEVEL":0,"ABMON_12":131097,"PTHREAD_KEYS_MAX":128,"EISCONN":106,"EMSCRIPTEN_FETCH_APPEND":8,"EMSCRIPTEN_FULLSCREEN_CANVAS_SCALE_NONE":0,"AUDIO_U16MSB":4112,"EMSCRIPTEN_FULLSCREEN_FILTERING_BILINEAR":2,"_SC_2_CHAR_TERM":95,"EMSCRIPTEN_EVENT_TOUCHEND":23,"_SC_AIO_LISTIO_MAX":23,"EMSCRIPTEN_FETCH_PERSIST_FILE":4,"_SC_BC_SCALE_MAX":38,"ENOTBLK":15,"EAI_BADFLAGS":-1,"EOVERFLOW":75,"EM_FUNC_SIG_VI":33554432,"SDL_AUDIO_MASK_SIGNED":32768,"EMSCRIPTEN_FULLSCREEN_FILTERING_NEAREST":1,"ABMON_11":131096,"ABMON_10":131095,"AT_FDCWD":-100,"EM_HTML5_SHORT_STRING_LEN_BYTES":32}}
```
Comment (thibault-martinez) [inline tools/emscripten/emscripten_cache/asmjs/generated_struct_info.json]: This shouldn't be pushed ?

File: tests/BUILD [hunk 7]
```diff
@@ -11,3 +11,16 @@ cc_library(
...
+
+emcc_binary(
+    name = "hi.js",
+    srcs = [
+        "hi.c",
+        "hi.h",
+    ],
+    #List of functions you would like to be able to wrap/cal from JS
+    #Please note that function name should be prefixed with _
+    linkopts = ["-s EXPORTED_FUNCTIONS='[\"_hi\"]'"],
```
Comment (thibault-martinez) [inline tests/BUILD]: This is mandatory ? We'll have to export all MAM API functions ?

File: tests/BUILD [hunk 8]
```diff
@@ -11,3 +11,16 @@ cc_library(
     #    includes = [".",],
     visibility = ["//visibility:public"],
 )
+
+load("//tools/emscripten:defs.bzl", "emcc_binary")
+
+emcc_binary(
+    name = "hi.js",
+    srcs = [
```
Comment (thibault-martinez) [inline tests/BUILD]: Does that work if we don't have `srcs` but only `deps` ?

File: tools/emscripten/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+##How to create a target
+Add "emcc_binary" rule defined in "defs.bzl" to create a "js" and "wasm" file
```
Comment (thibault-martinez) [inline tools/emscripten/README.md]:
```suggestion
Add `emcc_binary` rule defined in `defs.bzl` to create a `.js` and `.wasm` file
```

File: tools/emscripten/README.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+##How to consume js files
+
+One option at least is to create one js module that wrap and export the functions 
+that was already exported via the emcc_binary (using linker flag "-s EXPORTED_FUNCTIONS=[]")
```
Comment (thibault-martinez) [inline tools/emscripten/README.md]:
```suggestion
that was already exported via the emcc_binary (using linker flag `-s EXPORTED_FUNCTIONS=[]`)
```

File: tools/emscripten/cc_toolchain_config.bzl [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        default_link_flags_feature,
+        dbg_feature,
+        fastbuild_feature,
+        opt_feature,
+        user_compile_flags_feature,
+        sysroot_feature,
+        unfiltered_compile_flags_feature,
+    ]
+
+    cxx_builtin_include_directories = [ctx.var["workspace"], ctx.var["output_base"]]
```
Comment (thibault-martinez) [inline tools/emscripten/cc_toolchain_config.bzl]: So this is to include which directories ?

File: tools/emscripten/emcc.sh [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  outbase="$(basename "$baseout")"
+  (
+      cd "$outdir";
+      sources="$outbase."?(html|js|wasm|mem|data|worker.js)
+      tar cvf $tarfile $sources
+  )
+fi
+
+
+# rm -r tmp
```
Comment (thibault-martinez) [inline tools/emscripten/emcc.sh]: Where is this directory created ? I usually have a `tmp` folder in my current directory, I wouldn't want it to be erased by this script

**Comment by oopsmonk** [comment]
The usage of `hi.js` could be in a simple file:
```javascript
'use strict'                                                                                                                                                                 

let Module = require('./path-to/hi.js')

Module.onRuntimeInitialized = _ => {
    const say_hi = Module.cwrap('hi', 'null', []);
    say_hi();
}
```

**Comment by thibault-martinez** [comment]
What about the files in the cache ? Do we really have to push them ? Can't they be generate the first time it's run ?

**Comment by oopsmonk** [comment]
> What about the files in the cache ? Do we really have to push them ? Can't they be generate the first time it's run ?

we can find `EM_CACHE` and build links to `emscripten_cache` of this repo on the system.

**Pull request merged**